terraform/deployments/cluster-services/aws_lb_controller.tf

```diff
@@ -0,0 +1,23 @@
+# aws_lb_controller.tf manages the in-cluster components of the AWS Load
+# Balancer Controller.
+#
+# The AWS IAM resources for the AWS LB Controller are in
+# ../cluster-infrastructure/aws_lb_controller_iam.tf. See
+# https://github.com/alphagov/govuk-infrastructure/blob/main/docs/architecture/decisions/0003-split-terraform-state-into-separate-aws-cluster-and-kubernetes-resource-phases.md#decision for rationale.
+
+resource "helm_release" "aws_lb_controller" {
+  name       = "aws-load-balancer-controller"
+  repository = "https://aws.github.io/eks-charts"
+  chart      = "aws-load-balancer-controller"
+  version    = "1.2.6" # TODO: Dependabot or equivalent so this doesn't get neglected.
+  namespace  = "kube-system"
```

Comment:
Forgive my ignorance, is `kube-system` appropriate for user-defined objects, or is this LB controller part of the Kubernetes system?

Reply:
@bilbof that is a good question, and the picture on that seems a bit unclear. The [Kubernetes docs](https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/#viewing-namespaces) state that`kube-system` is intended for objects created by Kubernetes itself, which implies that we shouldn't deploy anything into it - but then [Amazon's docs](https://docs.aws.amazon.com/eks/latest/userguide/aws-load-balancer-controller.html) say to deploy it to `kube-system`, as do the [AWS LB Controller docs](https://kubernetes-sigs.github.io/aws-load-balancer-controller/v2.2/deploy/installation/).

So, not clear at all really, and I think you can argue it either way, but right now I'd say that we should use a separate namespace for supporting services like this, `cluster-services` or similar, as we'll have other components like Prometheus, Fluentbit etc. which should logically be organised into the same NS which don't seem to naturally belong in `kube-system` - that way we should avoid any unforeseen consequences of having third-party objects in `kube-system`, and allow us to apply RBAC controls to the `cluster-services` namespace if necessary, which might be a problem with `kube-system`.

Reply:
That makes sense, thanks for explaining!

Reply:
Great question. kube-system is basically for cluster API machinery. Objects which are managed by Kubernetes itself (i.e. the master node) certainly belong in there (for example coredns, kube-proxy, cloud provider's CNI plugin and so on). Deploying stuff into kube-system is definitely not something you'd do as a user of the cluster. In this case it's a bit of a fuzzy area because the LB controller is a basic service that's pretty much an essential part of the cluster, and it's the cloud provider's supported ingress controller addon. Both the [maintainer SIG's install guide](https://kubernetes-sigs.github.io/aws-load-balancer-controller/v2.2/deploy/installation/) and [Amazon's](https://docs.aws.amazon.com/eks/latest/userguide/aws-load-balancer-controller.html) say to use `kube-system`, and it seems reasonable in this case - so we went with that.

Reply:
I'll be surprised if we end up with anything else which warrants going in kube-system, btw.

Reply:
FWIW I can't think of any way that it could cause any practical problem by being in kube-system. I'd probably just be guided by the principle of least surprise on this one, so both manuals suggesting kube-system suggests that's a reasonable choice (and it's the first place I'd look if I wanted to grab the logs for the cloud provider's ingress controller). 

If there are strong feelings about this then I'm happy to bung it in its own namespace and re-test. It'd be another thing for people to have to discover if they're debugging though (because all the offical docs have kube-system in their examples).

Reply:
SGTM, I'm happy for us to follow the AWS docs and keep this in the `kube-system` namespace.

Reply:
No, no strong feelings on this from me - as I say, this can clearly be argued either way. So yeah, stick with `kube-system` - if there are any unintended consequences then it should be trivial to re-deploy to a different namespace.

Reply:
The equivalent controller in GKE lives in kube-system FWIW (as did all our official cluster addons), though those were all properly managed by the product — we didn't make the customers install them themselves with a handy [30+ step manual process](https://docs.aws.amazon.com/eks/latest/userguide/aws-load-balancer-controller.html) 😛).